### DISCUSSION TITLE
How to define an edge marker in a custom Edge?

### DISCUSSION THREAD

**Discussion by wowtah** [xyflow/xyflow#4300] [Q&A] [answered]
I created a custom Edge. And whenever this edge type is used I want to always show the same endMarker. This way I don't have to define it for every Edge in my edge collection. 

So I thought something like:

```
import React from "react";
import { BaseEdge, EdgeLabelRenderer, EdgeMarkerType, EdgeProps, getBezierPath, MarkerType } from "reactflow";

export default function CustomEdge({
  id,
  sourceX,
  sourceY,
  targetX,
  targetY,
  sourcePosition,
  targetPosition,
  style = {},

}: EdgeProps) {
  const [edgePath, labelX, labelY] = getBezierPath({
    sourceX,
    sourceY,
    sourcePosition,
    targetX,
    targetY,
    targetPosition,
  });

  const customMarkerEnd: EdgeMarkerType = {
    type: MarkerType.ArrowClosed,
    width: 20,
    height: 20,
    color: "#FF0072",
  };

  return (
    <>
      <BaseEdge path={edgePath} markerEnd={customMarkerEnd} />
      <EdgeLabelRenderer>
        <div
          style={{
            position: "absolute",
            transform: `translate(-50%, -50%) translate(${labelX}px,${labelY}px)`,
            fontSize: 12,
            pointerEvents: "all",
          }}
          className="nodrag nopan"
        >
          My custom label renderer
        </div>
      </EdgeLabelRenderer>
    </>
  );
}

```

But BaseEdge expects a string for markerEnd, so this doesn't fly. 

**How do I achieve this?** 

I would also like to put fixed styling for the BaseEdge in my custom Edge. I can't get this to work (/show) also. 

Any help would be appreciated :)

**Comment by bcakmakoglu** [comment, marked as answer]
You can use your *own* marker definitions in edges.

[Check out the edge markers example.](https://reactflow.dev/examples/edges/markers)
The example shows you, for example, how to use a custom edge marker like this:
<img width="181" alt="image" src="https://github.com/xyflow/xyflow/assets/78412429/f1ba0fa6-927f-4155-9806-9570258fb76f">

Of course you don't need to use some logo image as a marker like the example does, you could basically just copy + paste the existing marker from the ReactFlow code and re-create your own marker with a fixed id.

```tsx
      <svg style={{ position: "absolute", top: 0, left: 0 }}>
        <defs>
          <marker
            className="react-flow__arrowhead"
            id="marker-pink"
            markerWidth="40"
            markerHeight="40"
            viewBox="-10 -10 20 20"
            markerUnits="strokeWidth"
            orient="auto-start-reverse"
            refX="0"
            refY="0"
          >
            <polyline
              style={{
                stroke: "hotpink",
                fill: "hotpink",
                strokeWidth: 1,
              }}
              strokeLinecap="round"
              strokeLinejoin="round"
              points="-5,-4 0,0 -5,4 -5,-4"
            />
          </marker>
        </defs>
      </svg>
      <ReactFlow defaultNodes={defaultNodes} defaultEdges={defaultEdges} />
```

Now you can reference that marker anywhere with `url('#marker-pink')` (`marker-pink` is just the `id` I chose for this example, you are free to choose whatever name you like).

```tsx
<BaseEdge path={edgePath} markerEnd="url('#marker-pink')" />
```

[You can find a working example right here.](https://codesandbox.io/p/sandbox/little-pine-wpvxhw?file=%2FCustomEdge.js%3A44%2C27)

**Comment by DaniloRezende** [comment]
Hi, I have a question about this. I understand the solution, but I'd like to make a self-contained custom marker in a modular way. That is, I don't want the rest of my app having to know details of my custom edge. How do we achieve this?

**Reply by bcakmakoglu** [reply]
You can move the custom marker definition into your custom edge component and reference the id from there, 
for example
```tsx
      <svg style={{ position: "absolute", top: 0, left: 0 }}>
        <defs>
          <marker
            className="react-flow__arrowhead"
            id={`${id}-marker-pink`}
            markerWidth="40"
            markerHeight="40"
            viewBox="-10 -10 20 20"
            markerUnits="strokeWidth"
            orient="auto-start-reverse"
            refX="0"
            refY="0"
          >
            <polyline
              style={{
                stroke: "hotpink",
                fill: "hotpink",
                strokeWidth: 1,
              }}
              strokeLinecap="round"
              strokeLinejoin="round"
              points="-5,-4 0,0 -5,4 -5,-4"
            />
          </marker>
        </defs>
      </svg>
      <BaseEdge path={edgePath} markerEnd={`url(#${id}-marker-pink)`} />
```

[Here's an example.](https://codesandbox.io/p/sandbox/jovial-thompson-sn5jxm?file=%2FCustomEdge.js%3A31%2C1-57%2C73)

**Reply by DaniloRezende** [reply]
I see. Thank you for the speedy reply!

**Reply by 3goats** [reply]
Hi do you have a Svelte Flow version of this example please ?
